### ISSUE TITLE
Proposal: JupyterLab Completer UI

### ISSUE THREAD

**Issue by edzkite** [jupyterlab/jupyterlab#7983]
[Kite](https://kite.com) is building a JupyterLab extension. As part of the [Completer API refactor](https://github.com/jupyterlab/jupyterlab/issues/7044#issuecomment-594249225), we’d also like to enhance the completer UI.

Current JupyterLab UI:
<img width="251" alt="jupyter" src="https://user-images.githubusercontent.com/47993402/75834088-e13f5a80-5d6f-11ea-8c4a-5982c34a796b.png">
Prospective New UI:
<img width="272" alt="kite" src="https://user-images.githubusercontent.com/47993402/75834091-e2708780-5d6f-11ea-9c56-b783d3d8b9b5.png">

The two changes in this proposal are as follows:

### Completion Source Indicator

Currently, there’s no way for completions from different providers to visually differentiate themselves. One option is to enhance the type display, with the option to render a SVG or PNG mask. This solution has the benefit of allowing providers to give visual descriptors about the completions they provide.

Another option is to enable providers to add a completion source to the type extended node, as is done in Kite’s VSCode extension:
<img width="523" alt="kite2" src="https://user-images.githubusercontent.com/47993402/75834175-1c418e00-5d70-11ea-91e4-ce4949fd37d7.png">

As noted in https://github.com/jupyterlab/jupyterlab/issues/7044#issuecomment-594249225, we would make this the one and only completer UI. However, the UI enhancements here are optional; if not set, the view should render as is before.

### Documentation Display

The second change is to enable display of completion item documentation along the bottom of the completion list. We could create a new child `documentationNode` at https://github.com/jupyterlab/jupyterlab/blob/master/packages/completer/src/widget.ts#L762.

The node would use a Markdown parser internally, which would enable interoperability between Markdown, HTML, and plain text.

**Comment by tgeorgeux** [comment]
@edzkite Thanks for bringing this up Ed, I'm not familiar exactly how completers work, would a person be running completions from multiple sources at the same time? ie: Kite completions, an LSP server for python developers, and Jupyterlab's built-in completions?

My follow-up question (if the answer above is yes) would be how much impact does it have on the user to know where the completions are coming from? I can imagine it might help users to see which completions they're using more often for instance. In short, I like the idea of giving users more control over their actions, but I don't like cluttering the screen whenever we can avoid it.

**Comment by edzkite** [comment]
That is correct. A JupyterLab user would be able to get completions from both the LSP server and Kite.

In this setup, it’s extremely likely that the two providers will send the same completions to the completions list. Without visual differentiation and de-duping, this could give off the impression that completions are buggy.

Even with de-duping, adding a visual indicator would give users more insight into the value add of their chosen extensions. If all completions looked the same, there’s no way for a user to know a given extension moved the needle, or that the extension is actually working.

**Comment by Juan-132** [comment]
Could we not simply require the user to make a choice, either Kite or LSP?
How much extra added value is expected from enabling them both at the same time?

**Comment by edzkite** [comment]
There may be some instances where one provider is able to surface completions while the other is not. Enabling both to run at the same time allows the best of both worlds.

In VSCode, if a user has the MS Python extension and Kite, both are used to generate completions as well.

**Comment by edzkite** [comment]
Closing as said changes have been merged.
